File: benchmarks/build_speed_acopf_nlp.jl
# Build-speed prototype (NLP, JuMP/Julia) — exact AC optimal power flow, polar form.
#
# Same model as benchmarks/build_speed_acopf_nlp.py, so JuMP build time is directly
# comparable to Pyomo — the head-to-head for the AC-OPF backbone (the two tools
# that can express general non-convex NLP). Times the BUILD; solves a small case
# with Ipopt for the correctness check.
#
# Setup (once), into the jumpbench project:  Pkg.add("Ipopt")
#
# Usage:
#   julia build_speed_acopf_nlp.jl            # build-time sweep
#   julia build_speed_acopf_nlp.jl check      # also solve small, print objective
#   julia build_speed_acopf_nlp.jl N          # custom single size (N lines)
import Pkg
Pkg.activate(get(ENV, "JUMPBENCH_PROJECT", joinpath(homedir(), "jumpbench")); io=devnull)

using JuMP
const HAS_IPOPT = try
    @eval import Ipopt
    true
catch
    false
end

function _data(N)
    r = fill(0.01, N); x = fill(0.03, N)
    g = r ./ (r .^ 2 .+ x .^ 2)
    b = -x ./ (r .^ 2 .+ x .^ 2)
    Pd = vcat(0.0, fill(0.10, N))            # bus 1 = slack (no load)
    Qd = vcat(0.0, fill(0.05, N))
    cost = [1.0 + 0.05 * (i - 1) for i in 1:(N + 1)]
    return g, b, Pd, Qd, cost
end

function ybus_feeder(g, b, N)
    n = N + 1
    Gd = zeros(n); Bd = zeros(n)
    for k in 1:N                              # line k between bus k and k+1
        Gd[k] += g[k]; Gd[k + 1] += g[k]
        Bd[k] += b[k]; Bd[k + 1] += b[k]
    end
    nbradm(i, j) = (k = min(i, j); (-g[k], -b[k]))
    return Gd, Bd, nbradm
end

function build_model(N, data; do_solve=false)
    g, b, Pd, Qd, cost = data
    n = N + 1
    Gd, Bd, nbradm = ybus_feeder(g, b, N)
    nbrs(i) = filter(j -> 1 <= j <= n, (i - 1, i + 1))
    m = Model()
    @variable(m, 0.9 <= V[1:n] <= 1.1, start = 1.0)
    @variable(m, th[1:n], start = 0.0)
    @variable(m, 0 <= Pg[1:n] <= 5)
    @variable(m, -3 <= Qg[1:n] <= 3)
    fix(V[1], 1.0; force = true)
    fix(th[1], 0.0; force = true)
    @constraint(m, pb[i = 1:n],
        Pg[i] - Pd[i] == V[i]^2 * Gd[i] +
        sum(V[i] * V[j] * (nbradm(i, j)[1] * cos(th[i] - th[j]) + nbradm(i, j)[2] * sin(th[i] - th[j]))
            for j in nbrs(i)))
    @constraint(m, qb[i = 1:n],
        Qg[i] - Qd[i] == -V[i]^2 * Bd[i] +
        sum(V[i] * V[j] * (nbradm(i, j)[1] * sin(th[i] - th[j]) - nbradm(i, j)[2] * cos(th[i] - th[j]))
            for j in nbrs(i)))
    @objective(m, Min, sum(cost[i] * Pg[i] for i in 1:n))
    if do_solve
        set_optimizer(m, Ipopt.Optimizer)
        set_silent(m)
        optimize!(m)
        return objective_value(m)
    end
    return nothing
end

function timed_build(N, data; repeats=2)
    best = Inf
    for _ in 1:repeats
        GC.gc()
        best = min(best, @elapsed build_model(N, data))
    end
    return best
end

function run_sweep(sizes)
    build_model(8, _data(8))   # warm up JIT
    println(rpad("builder", 10), lpad("N lines", 8), lpad("cons", 8), lpad("build_s", 10))
    for N in sizes
        println(rpad("jump", 10), lpad(N, 8), lpad(2 * (N + 1), 8),
                lpad(round(timed_build(N, _data(N)), digits=3), 10))
    end
end

function correctness()
    obj = build_model(8, _data(8); do_solve=true)
    println("\nCorrectness (small AC OPF NLP solve, N=8, JuMP+Ipopt):")
    println("  jump   obj=", obj, "   (Pyomo should match)")
end

nums = [parse(Int, a) for a in ARGS if occursin(r"^\d+$", a)]
run_sweep(length(nums) >= 1 ? [nums[1]] : [100, 1000, 10000])
"check" in ARGS && correctness()

File: benchmarks/build_speed_acopf_nlp.py
"""Build-speed prototype (NLP): exact AC optimal power flow, polar form.

This is the *non-convex* class that the roadmap's exact unbalanced AC OPF needs.
Only Pyomo and JuMP can express general NLP (linopy/pyoframe are linear, CVXPY is
convex-only), so this is the Pyomo-vs-JuMP head-to-head for the AC-OPF backbone.
JuMP version: build_speed_acopf_nlp.jl. Times the model BUILD; solves a small case
with Ipopt for the correctness check.

Model: standard polar-form AC OPF on a radial feeder of N+1 buses (bus 0 = slack,
V0=1, theta0=0). Series admittance per line from (r, x); a generator and a load at
each bus. The non-convexity is the bus power-injection equations:

    P_i = V_i^2 G_ii + sum_{j~i} V_i V_j (G_ij cos(t_i-t_j) + B_ij sin(t_i-t_j))
    Q_i = -V_i^2 B_ii + sum_{j~i} V_i V_j (G_ij sin(t_i-t_j) - B_ij cos(t_i-t_j))
    Pg_i - Pd_i = P_i,   Qg_i - Qd_i = Q_i

minimise sum_i c_i Pg_i, with voltage and generation bounds.

Usage:
    python benchmarks/build_speed_acopf_nlp.py --check
    python benchmarks/build_speed_acopf_nlp.py --repeats 2
"""
from __future__ import annotations

import argparse
import time

import numpy as np


def _data(N):
    r = np.full(N, 0.01)
    x = np.full(N, 0.03)
    g = r / (r ** 2 + x ** 2)          # series conductance per line
    b = -x / (r ** 2 + x ** 2)         # series susceptance per line
    Pd = np.concatenate([[0.0], np.full(N, 0.10)])   # load per bus 0..N
    Qd = np.concatenate([[0.0], np.full(N, 0.05)])
    cost = 1.0 + 0.05 * np.arange(N + 1)             # gen cost per bus (cheapest at slack)
    return g, b, Pd, Qd, cost


def _ybus_feeder(g, b, N):
    """Tridiagonal Ybus of the feeder. Returns Gd,Bd (diagonals, len N+1) and a
    function nbr_adm(i,j)->(Gij,Bij) for adjacent buses."""
    Gd = np.zeros(N + 1)
    Bd = np.zeros(N + 1)
    for k in range(1, N + 1):           # line k between bus k-1 and k
        for i in (k - 1, k):
            Gd[i] += g[k - 1]
            Bd[i] += b[k - 1]

    def nbr_adm(i, j):
        k = max(i, j)                   # the line index between i and j
        return -g[k - 1], -b[k - 1]
    return Gd, Bd, nbr_adm


def build_pyomo(N, data, solve=False):
    import pyomo.environ as pyo
    g, b, Pd, Qd, cost = data
    Gd, Bd, nbr_adm = _ybus_feeder(g, b, N)
    t0 = time.perf_counter()
    m = pyo.ConcreteModel()
    m.bus = pyo.RangeSet(0, N)
    m.V = pyo.Var(m.bus, bounds=(0.9, 1.1), initialize=1.0)
    m.th = pyo.Var(m.bus, initialize=0.0)
    m.Pg = pyo.Var(m.bus, bounds=(0.0, 5.0), initialize=0.0)
    m.Qg = pyo.Var(m.bus, bounds=(-3.0, 3.0), initialize=0.0)
    m.V[0].fix(1.0)
    m.th[0].fix(0.0)

    def nbrs(i):
        return [j for j in (i - 1, i + 1) if 0 <= j <= N]

    def pbal(m, i):
        inj = m.V[i] ** 2 * Gd[i]
        for j in nbrs(i):
            Gij, Bij = nbr_adm(i, j)
            inj += m.V[i] * m.V[j] * (Gij * pyo.cos(m.th[i] - m.th[j]) + Bij * pyo.sin(m.th[i] - m.th[j]))
        return m.Pg[i] - Pd[i] == inj

    def qbal(m, i):
        inj = -m.V[i] ** 2 * Bd[i]
        for j in nbrs(i):
            Gij, Bij = nbr_adm(i, j)
            inj += m.V[i] * m.V[j] * (Gij * pyo.sin(m.th[i] - m.th[j]) - Bij * pyo.cos(m.th[i] - m.th[j]))
        return m.Qg[i] - Qd[i] == inj

    m.pbal = pyo.Constraint(m.bus, rule=pbal)
    m.qbal = pyo.Constraint(m.bus, rule=qbal)
    m.obj = pyo.Objective(expr=sum(cost[i] * m.Pg[i] for i in m.bus))
    build = time.perf_counter() - t0
    obj = None
    if solve:
        pyo.SolverFactory("ipopt").solve(m)
        obj = float(pyo.value(m.obj))
    return build, 2 * (N + 1), obj


BUILDERS = {"pyomo": build_pyomo}


def run_sweep(sizes, repeats=1):
    print(f"{'builder':10s} {'N buses':>8s} {'cons':>8s} {'build_s':>9s}")
    for N in sizes:
        data = _data(N)
        best = None
        for _ in range(repeats):
            bld, ncon, _ = build_pyomo(N, data)
            best = bld if best is None else min(best, bld)
        print(f"{'pyomo':10s} {N:8d} {ncon:8d} {best:9.3f}")


def correctness_check():
    N = 8
    _, _, obj = build_pyomo(N, _data(N), solve=True)
    print(f"\nCorrectness (small AC OPF NLP solve, n={N}):")
    print(f"  pyomo   obj={obj}")
    return obj


def main(argv=None):
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("--repeats", type=int, default=1)
    p.add_argument("--check", action="store_true")
    p.add_argument("--n", type=int, default=None)
    args = p.parse_args(argv)
    sizes = [args.n] if args.n else [100, 1000, 10000]
    run_sweep(sizes, repeats=args.repeats)
    if args.check:
        correctness_check()


if __name__ == "__main__":
    main()
